tests/gis_tests/layermap/tests.py

```diff
@@ -319,7 +319,8 @@ def db_for_write(self, model, **hints):
         return self.db_for_read(model, **hints)
 
     def allow_relation(self, obj1, obj2, **hints):
-        return None
+        # Required to be True after fixing bug #27629
+        return True
```

Comment:
Because allow_relation is now called more often (correctly), I'm worried that improperly configured routers that were never fully exercised before may start to cause failures in existing systems that use Django.

Reply:
may be adding some notes about that on release note?

Reply:
The documentation says that, `allow_relation()` should "return True if a relation between obj1 and obj2 should be allowed, False if the relation should be prevented, or None if the router has no opinion." I didn't investigate but I can't logically see how a correct implementation would require router that previously had no opinion to now return True.

Reply:
The problem is that the documentation is not clear what should happen if the router returns `None`. Are relationships by default allowed or not allowed?

The current and "fixed" implementation checks if a relationship is allowed via `if not router.allow_relation(...)`. See `related_descriptors.py` line 211 and 457 (old line numbers). A return value of `None` would mean that a relationship is not allowed (i.e., it must explicitly be allowed by returning `True`). I kept this behavior.

However, since `allow_relation()` was not being called **at all** in the specific scenario here (hence my pull request), this issue was never caught. If the proper conditions were met such that `allow_relation()` were to be called (such as re-assigning the foreign key after the initial save), then the relationship would never have been allowed in the first place.

Reply:
Actually, `None` never gets returned to the code in `related_descriptors.py`...

The [current router behavior (utils.py)](https://github.com/django/django/blob/84dcd1624784c2239f96a97b3151145a85dfbbe3/django/db/utils.py#L281), if all the routers return None, is to disallow cross-database relationships.

The problem with this test is that the `allow_relation()` function here was never called, until now.